morpheus/algorithms/impl/dynamic_matrix/copy_impl.hpp

```diff
@@ -0,0 +1,81 @@
+/**
+ * copy_impl.hpp
+ *
+ * EPCC, The University of Edinburgh
+ *
+ * (c) 2021 The University of Edinburgh
+ *
+ * Contributing Authors:
+ * Christodoulos Stylianou (c.stylianou@ed.ac.uk)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * 	http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef MORPHEUS_ALGORITHMS_IMPL_DYNAMIC_MATRIX_COPY_IMPL_HPP
+#define MORPHEUS_ALGORITHMS_IMPL_DYNAMIC_MATRIX_COPY_IMPL_HPP
+
+#include <morpheus/core/matrix_tags.hpp>
+#include <morpheus/core/exceptions.hpp>
+#include <morpheus/containers/dynamic_matrix.hpp>
+
+namespace Morpheus {
+// forward decl
+template <typename SourceType, typename DestinationType>
+void copy(const SourceType& src, DestinationType& dst);
+
+template <typename SourceType, typename DestinationType>
+void copy(const SourceType& src, SourceType& dst);
+
+namespace Impl {
+
+struct copy_fn {
+  using result_type = void;
+
+  template <typename SourceType, typename DestinationType>
+  result_type operator()(const SourceType& src, SourceType& dst) const {
+    Morpheus::copy(src, dst);
+  }
+
+  template <typename SourceType, typename DestinationType>
+  result_type operator()(const SourceType& src, DestinationType& dst) const {
+    std::string msg("Invalid use of the copy interface: ");
+    throw Morpheus::RuntimeException(msg + src.name() + " " + dst.name() +
+                                     "\n");
+  }
+};
+
+template <typename SourceType, typename DestinationType>
+void copy(const SourceType& src, DestinationType& dst, Morpheus::DynamicTag,
+          Morpheus::Impl::SparseMatTag) {
+  std::visit(std::bind(Impl::copy_fn(), std::placeholders::_1, std::ref(dst)),
```

Comment:
All possible combinations for SourceType and DestinationType must be provided for copy_fn().
Only the same types should overload( eg Morpheus::copy(Coo, Coo)) otherwise a run-time exception will be generated (eg for Morpheus::copy(Coo, Csr)).
The way this is currently implemented all the overloads choose the copy_fn() overload that throws.